### ISSUE TITLE
_InputStream should expose seek method so that we can read environ["wsgi.input"] more than once

### ISSUE THREAD

**Issue by Wolfcast** [twisted/twisted#11875]
In our application we do `environ["wsgi.input"].read()` to parse the request payload. In some case we then rewrite the request and on some others we leave the request as is. When we leave it as-is we will need to access `environ["wsgi.input"]` later down the line but since the cursor is at the end of the stream when we read it again it's empty. We need to be able to place the cursor back at the beginning like this: `environ["wsgi.input"].seek(0)`.

I already implemented the solution and it works (PR incoming). It should not have any impact since it's a new method that is exposed.

**Comment by adiroiban** [comment]
I don't have experience with WSGI.

I know that for generic Twisted HTTP request content, it's best to have real streaming input.

That is, just read with pause and without any seek.

Imagine receiving a 10GB request (maybe it makes no sense for WSGI)
but for this case you just receive the data once and then it's up to you to store it somewhere.

If you need to see it later, make sure you create a separate copy and then seek on that copy.

just my feedback.

**Comment by Wolfcast** [comment]
@adiroiban Yes it was our first workaround but copying the content of `environ["wsgi.input"]` will double RAM usage for a short time (it can be nothing but it can be significant for large requests). We have thousand of concurrent users and the easiest/cleanest solution/fastest (IMO) is the `seek(0)` solution. I understand that what we do is somewhat out of the ordinary (rewrite requests in some cases) and the added feature won't be used but almost no one but my target is a production app which we can't change too much without freaking out the QA department :)

**Comment by adiroiban** [comment]
I don't know what is best, since I have zero experience with WSGI applications.

I can see that `_InputStream` is defined as a private class and `_WSGIResponse` is also private.

I guess that option is to allow for a WSGI server to implement a custom InputStream.

------------

For the HTTP protocol, I have a custom request factory that I using for the HTTP site,
and this is used to create custom request content handling. 

------------

So maybe for WSGI, we can have something like `WSGIResource.requestContentFactory` or call it just `WSGIResource.inputStreamFactory` that is used to return a custom `_InputStream`

-----

Again, just some very high level API.

Maybe WSGI standard requires seekable streams, and in that case, this is a bug in Twisted and not a new feature.

I can see that twisted.web.http.Request.content has no restriction in terms of seek operations.

And I guess that all file-like objects have `seek()` 

so your suggestion might be valid

```
    @ivar content: A file-like object giving the request body.  This may be
        a file on disk, an L{io.BytesIO}, or some other type.  The
        implementation is free to decide on a per-request basis.
    @type content: L{typing.BinaryIO}

```

Thanks again for your feedback.

Feel free to send a PR and I hope that someone else can help with the review.

Regards
